byterun/meta.c

```diff
@@ -49,40 +50,84 @@ CAMLprim value caml_get_section_table(value unit)
                                      caml_section_table_size);
 }
 
-CAMLprim value caml_reify_bytecode(value prog, value len)
+struct bytecode {
+  code_t prog;
+  asize_t len;
+};
+#define Bytecode_val(p) ((struct bytecode*)Data_abstract_val(p))
+
+CAMLprim value caml_reify_bytecode(value ls_prog, value debuginfo, value digest_opt)
 {
+  CAMLparam3(ls_prog, debuginfo, digest_opt);
+  CAMLlocal4(clos, bytecode, retval, s);
   struct code_fragment * cf = caml_stat_alloc(sizeof(struct code_fragment));
-  value clos;
+  code_t prog;
+  asize_t len, off;
+  int i;
 
+  /* Convert from LongString.t to contiguous buffer */
+  len = 0;
+  for (i = 0; i < Wosize_val(ls_prog); i++) {
+    s = Field(ls_prog, i);
+    len += caml_string_length(s);
+  }
+
+  prog = (code_t)caml_stat_alloc(len);
+  off = 0;
+  for (i = 0; i < Wosize_val(ls_prog); i++) {
+    size_t s_len;
+    s = Field(ls_prog, i);
+    s_len = caml_string_length(s);
+    memcpy((char*)prog + off, Bytes_val(s), s_len);
+    off += s_len;
+  }
```

Comment:
Could you move this "convert from LongString.t to contiguous buffer" function to an auxiliary function, maybe in byterun/str.c? It would make the code easier to read, avoid mixing abstraction level, and it might be useful for someone else.

Reply:
Done, but I left the function in `meta.c`. The `LongString.t` type seems to be quite specific to the bytecode implementation, and unlikely to be of much use to other code which would probably use bigstrings instead. (Eventually, it should probably be replaced with bigstrings here too, pending perf worries).